core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/tracker/HttpRequest.scala

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2021-2022 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.eventgen.tracker
+
+import com.snowplowanalytics.snowplow.eventgen.collector.Api
+import com.snowplowanalytics.snowplow.eventgen.collector.Api._
+import com.snowplowanalytics.snowplow.eventgen.primitives.Url.genOriginHost
+import com.snowplowanalytics.snowplow.eventgen.primitives.{Url, genIp, genString, genUserAgent}
+import com.snowplowanalytics.snowplow.eventgen.tracker.HttpRequest.Method
+import org.scalacheck.Gen
+
+import java.time.Instant
+
+final case class HttpRequest(method: Method, headers: Map[String, String], qs: Option[HttpRequestQuerystring], body: Option[HttpRequestBody])
+
+object HttpRequest {
+  sealed trait Method {
+    val path: Api
+  }
+
+  object Method {
+    final case class Post(path: Api) extends Method
+    final case class Get(path: Api) extends Method
+    final case class Head(path: Api) extends Method
+
+    def gen: Gen[Method] = Gen.oneOf(genPost, genGet, genHead)
+
+    private def genPost: Gen[Method.Post] = Gen.oneOf(genApi(0), genApi(200)).map(Method.Post)
+    private def genGet: Gen[Method.Get] = Gen.oneOf(fixedApis, genApi(0), genApi(1)).map(Method.Get)
+    private def genHead: Gen[Method.Head] = Gen.oneOf(fixedApis, genApi(0), genApi(1)).map(Method.Head)
+  }
+
+  object Headers {
```

Comment:
I think using the `Headers` from the `HttpRequest` is an easy adjustment. I'll do that.

---

I believe there's a challenge with the idea of `CollectorPayload` being the top-level generator, from which both `HttpRequest` and `SdkEvent` are generated.

A couple of examples:

1.) We need to be able to generate http requests with either querystring or body, or neither, or both. All of these are valid options for which the collector has defined behaviours that we want to test.

Let's say we make `CollectorPayload.payload` an `Option[List[Body]]` and we add `CollectorPayload.querystring`, which is also optional.

We'll have to be able to generate come `CollectorPayload`s with both of these set to `None`, so that we can then derive an http request where both are missing. But then we won't be able to use these same `CollectorPayload`s to also generate `SdkEvent`, because they don't contain any actual event data.

2.) We need to be able to generate http requests with random paths, ie `Api(random, random)`. 

We can create `CollectorPayload` with random `Api` and we can use it to derive http requests. But again, we won't be able to create `SdkEvent` from it, because there's no enrich adaptor that handles the `random/random` path.

There are three options we have to solve the above:
- introduce bad rows, so that we can create failed events from the collector payloads described in the examples;
- introduce some mapping from 'incomplete' (missing body and qs) / 'faulty' (having random path) collector payload to 'correct' one before it is used to create an `SdkEvent`
- introduce some additional config, so we can generate `CollectorPayload` -> `HttpRequest` or `CollectorPayload` -> `SdkEvent`, but not `CollectorPayload` -> (`HttpRequest`, `SdkEvent`).

I think all three could be valid. But at this time, we don't actually need to make the decision which one is best. Because we already have everything we need to generate data and write tests.

---

I think also that `httpSink` is a bit pointless. What would the output look like? The data that the event generator creates is not a genuine http request that can be read and sent with curl or some other tool. It's just a wrapper around some building blocks for real http requests.

Any thoughts?

Reply:
1) Collector payload already [handles](https://github.com/snowplow-incubator/snowplow-event-generator/blob/main/core/src/main/scala/com/snowplowanalytics/snowplow/eventgen/collector/CollectorPayload.scala#L58:L69) both GET and POST style requests. It will probably need some extra code to integrate with the new httpRequest, but it shouldn't be much.

2) 
    - It makes sense to make HTTP request the bottom-level entity. The reason I suggested the opposite is it would be somewhat less code to refactor. But considering bad events http-request should be at the bottom.

    - Bad events are a separate issue, which probably deserves a separate PR. They could be generated by the mapping functions  `ColPayloadFromHttpRequest` and  `SdkEventFromColPayload`. It would require adding an AST to distinguish event types and refactoring the sinks and a bit of generation logic. To limit the scope sink could just drop the bad events, in the first release. 
    It would require a fair amount of config changes.

3) `httpSink` would generate genuine HTTP requests. The existing fs2 sink interface will work nicely with http4s client.

Reply:
@voropaevp Sorry, I don't understand. How would that solve the problems I listed in my previous comment about 'empty' requests (with no body and no qs) and about randomised paths?

Reply:
My comments ensure that the overall architecture of the application doesn't drift without very good reason. 
- Http client is implemented as Sink
- Generated events and http requests generated in pairs, so they could be verified.

Both problems you are solving would have to be solved with new code. The difference is integrating it into the existing event generation flow or building a new one.